crates/core/src/client/inner/cookie_store.rs

```diff
@@ -0,0 +1,151 @@
+use std::sync::Arc;
+
+use log::{error, warn};
+use reqwest::{cookie::CookieStore, header::HeaderValue, Url};
+
+use crate::callbacks::SecurePersistentStore;
+
+const COOKIE_STORE_KEY: &str = "COOKIE_CACHE";
+
+pub struct PersistentCookieStore {
+    store: Arc<reqwest_cookie_store::CookieStoreMutex>,
+    persistent_store: Option<Arc<dyn SecurePersistentStore>>,
+}
+
+impl PersistentCookieStore {
+    pub fn new(persistent_store: Option<Arc<dyn SecurePersistentStore>>) -> Self {
+        let cookie_store = if let Some(store) = &persistent_store {
+            if let Some(binary_json) = store.get(COOKIE_STORE_KEY.to_owned()) {
+                match cookie_store::serde::json::load(binary_json.as_slice()) {
+                    Ok(store) => store,
+                    Err(e) => {
+                        error!(
+                            "Failed to load cookie store: {} - defaulting to empty store",
+                            e
+                        );
+                        reqwest_cookie_store::CookieStore::default()
+                    }
+                }
+            } else {
+                reqwest_cookie_store::CookieStore::default()
+            }
+        } else {
+            warn!("No persistent store configured, no cookies will be loaded from disk.");
+            reqwest_cookie_store::CookieStore::default()
+        };
+
+        let store = Arc::new(reqwest_cookie_store::CookieStoreMutex::new(cookie_store));
+
+        Self {
+            store,
+            persistent_store,
+        }
+    }
+    pub fn get_cookie_list(&self, url: &Url) -> Option<Vec<String>> {
+        self.cookies(url).map(|header| {
+            header
+                .to_str()
+                .unwrap_or_default()
+                .split("; ")
+                .map(|s| s.to_string())
+                .collect::<Vec<String>>()
+        })
+    }
+
+    pub fn save(&self) {
+        let Some(store) = &self.persistent_store else {
+            warn!("No persistence provider while attempting to save, Cookies will not persist");
+            return;
+        };
+
+        let mut buffer = Vec::new();
+        let store_guard = self.store.lock().unwrap();
```

Comment:
Hmm. Given that you call this from an `impl Drop`, you might not wanna have an `unwrap`.

Reply:
that is fair, I'm going to move this out of drop. I thought I was being clever by making this run on clean up - but we pretty much have to run it after every insertion just in case something goes wrong anyways.